README.md
# Reading List
[![Bower version](https://badge.fury.io/bo/reading-list.svg)](http://badge.fury.io/bo/reading-list)

Why navigate when you don't have to?


## Setup

Include the CSS and JS you need to make this run:
```html
<link rel="stylesheet" type="text/css" href="jquery.reading-list.css">
<script src="jquery.reading-list.js"></script>
```

Make your HTML like this:
```html
<div class="reading-list-left-nav"> <!-- Minimap container -->
  <ul class="reading-list-mini-map">
    <li class="reading-list-mini-map-item" data-item-ref="article-1-slug">
      <span class="nav-viewing-indicator">&#x2713;</span>
      <a href="/article/article-1-slug">My First Article!</a>
    </li>
    <li class="reading-list-mini-map-item" data-item-ref="article-2-slug">
      <span class="nav-viewing-indicator">&#x2713;</span>
      <a href="/article/article-2-slug">My SECOND Article!</a>
    </li>
    <li class="reading-list-mini-map-item" data-item-ref="article-3-slug">
      <span class="nav-viewing-indicator">&#x2713;</span>
      <a href="/article/article-3-slug">This Is The Last Article :(!</a>
    </li>
  </ul>
</div>
<div class="reading-list-content"> <!-- Reading list initialized on this element -->
  <div class="reading-list-items">
      <div id="article-1-slug" class="reading-list-item" data-href="/article/article-1-slug">
        <div class="reading-list-loader">Up next My First Article!...</div>
      </div>
      <div id="article-2-slug" class="reading-list-item" data-href="/article/article-2-slug">
        <div class="reading-list-loader">Up next My SECOND Article!...</div>
      </div>
      <div id="article-3-slug" class="reading-list-item" data-href="/article/article-3-slug">
        <div class="reading-list-loader">Up next This Is The Last Article :(!...</div>
      </div>
  </div>
</div>
```

Then make your JS like this:
```javascript
$(document).on('ready', function () {
  // initialize reading list
  $('.reading-list-content').readingList({
    loadingThreshold: 300,
    lookingThresholdTop: 200,
    lookingThresholdBottom: 250,
    dataRetrievalSuccess: function ($item, data) { return data; },
    dataRetrievalFail: function ($item) { return 'fail!'; }
  });
});
```

As the user scrolls down, article content will be loaded according to the ```data-href```
attribute on each reading list item. Items are marked in the mini-map as active
based on the id of the reading list item which should match with the data-item-ref.

Whatever contents are inside of the reading list item, which in this example case is
```html
<div class="reading-list-loader">Some text...</div>
```
will be replaced by the content returned by the AJAX call.

### Options
Following is a breakdown of the options available for this plugin:

| Option | Notes |
|--------:|:-------|
| `addContent` | Define this content to add content to the end of the reading list when there are no more items to load. Expected to return a promise that will resolve with the content to append to the end of the list. |
| `dataRetrievalFail` | Reading list data failure callback. Non-falsy return values from this function will replace the contents of the reading list item. In the ideal case, this function will return HTML of some error message. |
| `dataRetrievalSuccess` | Reading list data transform callback to change received data to HTML. Non-falsy return values from this function will replace the contents of the reading list item. In the ideal case, this function will return the HTML of the reading list item. |
| `eventingThrottle` | Time in ms to delay eventing execution on scroll. Use larger values when considering slower browsers so that scrolling isn't affected by event triggering. |
| `isMobile` | A value that determines if the user is currently on a mobile device, used for changing reading list behaviors depending on how the reading list is being viewed. Either a function that returns a boolean, or a constant boolean value. |
| `loadingThreshold` | Height from the bottom of scrolling container to start loading. Either a function that returns a number or a constant number value. |
| `lookingThresholdBottom` | Bottom boundary of "looking" area, measured from top of window. Either a function that returns a number or a constant number value. |
| `lookingThresholdTop` | Top boundary of "looking" area, measured from top of window. Either a function that returns a number or a constant number value. |
| `miniMapActiveClass` | Class to apply to active minimap items. |
| `noEventBubbling` | Set to true to stop events from bubbling up the DOM tree, in which case, any event listeners must attach to the reading list element itself. This **must** be set to true for any reading lists that are nested inside another reading list! |
| `onPreSetup` | Set to a function that will fire before reading list is setup and initial eventing and loading is done. Reading list object will be passed in as first argument. |
| `onReady` | Set to a function that will fire when the reading list is ready and initial eventing and loading has been done. Reading list object will be passed in as first argument. |
| `scrollContainer` | Container that is used for scrolling. Customize this only if using a non-element--such as window, body, document, html--for scrolling. |
| `scrollToAddPx` | Number, or function that returns a number, to add to scroll to distance. Useful if some absolute position element (such as a sticky header) needs to be accounted for when scrolling to an item. |
| `scrollToSpeed` | Time in ms for scroll to event when scrolling to an article. |
| `selectorsItems` | Direct children of `itemsContainer`, the class used for individual reading list items. |
| `selectorsItemsContainer` | Direct child of element used to create reading list. |
| `selectorsItemsPreLoaded` | Filter selector for `selectorsItems` that distinguishes items that have already been loaded into the reading list before reading list initialization. |
| `selectorsMiniMapItems` | Selector for mini-map elements. Minimap elements can be anywhere on the page as long as they can be selected with this selector and have the `data-item-ref` attribute. If you are using multiple reading lists on a single page, this selector should be customized for each individual instance of a reading list. |

Note, that any styles that come along with the plugin must be applied to the corresponding customized classes here.

### Events
Following are events you can hook into as the page scrolls. Note: all events are triggered on the container used to build the reading list, so in the example, it would be on the ```#readingList``` element.

#### Reading List Events
Events applicable to the state of the entire reading list. No arguments are given
to event callbacks.

| Name | Notes |
|-----:|:----------|:------|
| ```reading-list-at-top``` | Fires when reading list hits the top. |
| ```reading-list-out-of-content``` | Reading list has run out of content. Internally, if ```addContent``` is defined it will be used when this event triggers to add new content to the end of the reading list. |

#### Item Events
Events applicable to individual reading list items. Note that:
1. All event callbacks are given the ```$item``` the event is for as the first
argument.
2. The ```Arguments``` listed are keys of the second argument into the callback.
3. ```callCount``` refers to the number of times the given event has been
triggered on that item.

| Name | Arguments | Notes |
|-----:|:----------|:------|
| ```reading-list-item-load-start``` | ```callCount``` ```direction``` | ```$item``` is starting to load, triggered by a scrolling ```direction```. |
| ```reading-list-item-load-done``` | ```callCount``` | ```$item``` is done loading. |
| ```reading-list-item-in-looking``` | ```callCount``` | ```$item``` has entered the looking area and is the active item. |
| ```reading-list-item-out-looking``` | ```callCount``` | ```$item``` has fallen out of the looking area and is no longer the active item. |
| ```reading-list-item-progress``` | ```progress``` | Has viewed ```progress``` ratio of the entire ```$item```. Use to update progress bars. To make ```progress``` a percentage, do ```progress * 100```.  |
| ```reading-list-start-scroll-to```| None |```$item``` is being scrolled into view.|
| ```reading-list-end-scroll-to```| None |```$item``` has been scrolled into view.|

## Usage Notes

### The "Looking" Area
Whatever the top item is in the reading list that falls within the "looking"
area is what is active and will be marked as such in the mini-map. For the
progress bar, the amount "read" is the distance from the top of the article to
the bottom of the "looking area".

### Preloading Items
Any reading list items that are loaded with the page should have the attribute ```data-load-status="loaded"``` to prevent reloading that item. In order to keep
styling consistent, the item should also get the ```loaded``` CSS class.

## Development

### Running the Example
```bash
$ npm run example
```

### Run tests
```bash
$ npm run test
```

### Creating a New Build
```bash
$ npm run build
```

## Running Tests/Development
```bash
npm install
bower install
npm test
```


CHANGELOG.md
## Version 0.2.0
- replaced `settings.scrollToAddPx` with `settings.scrollToSubtractPx`
  The value is now subtracted.

## Version 0.1.0

- using `$.fn.offset()` in favor of `$.fn.position()` to
  calculate element positions

- requires change to reading list implementations

```
SomeReadingList.prototype.scrollToAddPx = function () {
  return - this.$header.outerHeight(true);
};
```

Should return a negative value to add to scrollTop that is equal to
any sticky top header elements heights.


example/js/reading-list.js
(function (window, $, IScroll) {
  'use strict';

  var $window = $(window);
  var $document = $(window.document);

  var windowHeight = function () {
    return window.innerHeight || window.document.documentElement.clientHeight;
  };

  /**
   * Check that some part of given element is in view + threshold.
   */
  var withinLoadingThreshold = function (el, threshold) {
    // window dimesions
    var top = -threshold;
    var bot = windowHeight() + threshold;

    // check if element bounding box is within window dimensions
    var elBounding = el.getBoundingClientRect();
    var overTop = elBounding.top < top && elBounding.bottom < top;
    var belowBot = elBounding.top > bot && elBounding.bottom > bot;

    // if the whole thing is not over the top, or the whole thing is not below the bot,
    //  some part of it must be in the view
    return !(overTop || belowBot);
  };

  /**
   * Determine if a user is "looking" at an item. User is "looking" at the item if
   *  it falls within the boundaries created by the top and bottom thresholds.
   */
  var withinLookingArea = function (el, topThreshold, bottomThreshold) {
    // get top and bottom thresholds
    var lineTop = topThreshold;
    var lineBot = bottomThreshold;

    // check if element is within "looking" area dimensions
    var elBounding = el.getBoundingClientRect();
    var overTop = elBounding.top < lineTop && elBounding.bottom < lineTop;
    var belowBot = elBounding.top > lineBot && elBounding.bottom > lineBot;

    // not over the top or bottom, so some part of this is within the "looking" area
    return !(overTop || belowBot);
  };

  $document.ready(function () {

    var LOADING_THRESHOLD = 300;

    // only one threshold is really necessary depending on algo, but use both
    var VIEWING_THRESHOLD_TOP = 200;
    var VIEWING_THRESHOLD_BOT = 250;

    var $readingListMiniMap = $('#readingListMiniMap');
    var $readingListMiniMapItems = $readingListMiniMap.find('.reading-list-mini-map-item');
    var $readingListContent = $('#readingListContent');
    var $readingListArticles = $('#readingListArticles');
    var $readingListItems = $readingListArticles.find('.reading-list-item');
    var $activeItem;

    if ($.browser.mobile) {
      // mobile browser, use IScroll
      (new IScroll('readingListContent', {
        useNativeScroll: true
      }));
    }

    /**
     * Reading list event framework.
     */
    var eventing = function () {
      var scrollTop = $readingListContent.scrollTop();
      if (scrollTop <= 0) {
        // we're at the top of the reading list
        $readingListContent.trigger('reading-list-at-top');
      } else if (scrollTop + windowHeight() >= $readingListArticles.height()) {
        // we're at the bottom of the reading list
        $readingListContent.trigger('reading-list-at-bottom');
      }

      // check if individual reading list items are in view
      var $nowActive;
      $readingListItems.each(function (i, item) {
        var $item = $(item);
        var inThreshold = withinLoadingThreshold(item, LOADING_THRESHOLD);
        if(inThreshold && !$item.hasClass('in-threshold')) {
          $item.addClass('in-threshold');
          $readingListContent.trigger('reading-list-item-in-view', $item);
        } else if (!inThreshold && $item.hasClass('in-threshold')){
          $item.removeClass('in-threshold');
          $readingListContent.trigger('reading-list-item-out-view', $item);
        }

        // mark the higher up item in the looking area as the one being looked at
        var inLooking = withinLookingArea(item, VIEWING_THRESHOLD_TOP, VIEWING_THRESHOLD_BOT);
        if(inLooking && !$nowActive) {
          // in looking area, and we haven't assigned a now active item yet
          if (!$item.is($activeItem)) {
            // this is not the currently active item, so we'll want to fire off events and things
            if ($activeItem) {
              // new item in looking area, set it to the active item
              $activeItem.removeClass('in-looking');
              $readingListContent.trigger('reading-list-item-out-looking', $activeItem);
            }
            // add looking class to active item, trigger event
            $item.addClass('in-looking');
            $readingListContent.trigger('reading-list-item-in-looking', $item);
          }
          $nowActive = $item;
        }
      });
      // found an active item, set it to the active item
      $activeItem = $nowActive;
    };

    $readingListContent.on('reading-list-item-in-looking', function (event, item) {
      var $item = $(item);
      if ($item.children('article').length > 0) {
        var id = $item.children('article').attr('id');
        var $miniMapItem = $readingListMiniMapItems.filter('[data-item-ref="' + id + '"]').addClass('active');
console.log('in looking', id);
      }
    });
    $readingListContent.on('reading-list-item-out-looking', function (event, item) {
      var $item = $(item);
      if ($item.children('article').length > 0) {
        var id = $item.children('article').attr('id');
        var $miniMapItem = $readingListMiniMapItems.filter('[data-item-ref="' + id + '"]').removeClass('active');
console.log('out looking', id);
      }
    });
    $readingListContent.on('reading-list-item-in-view', function (event, item) {
      var $item = $(item);
      var name;
      if ($item.children('article').length > 0) {
        name = $item.children('article').attr('id');
      } else {
        name = 'adspace';
      }
    });
    $readingListContent.on('reading-list-item-out-view', function (event, item) {
      var $item = $(item);
      var name;
      if ($item.children('article').length > 0) {
        name = $item.children('article').attr('id');
      } else {
        name = 'adspace';
      }
    });
    $readingListContent.on('reading-list-at-top', function () {
      console.log('topped', $readingListContent.scrollTop());
    });
    $readingListContent.on('reading-list-at-bottom', function () {
      console.log('bottomed', $readingListContent.scrollTop());
    });

    eventing();
    $readingListContent.on('scroll', function () {
// TODO : might want to do some kind of debounce here
      eventing();
    });
    $readingListContent.on('resize', function () {
      eventing();
    });

  });

})(window, $, IScroll);
